atlasdb-coordination-impl/src/main/java/com/palantir/atlasdb/coordination/keyvalue/KeyValueServiceCoordinationStore.java

```diff
@@ -125,18 +125,45 @@ private KeyValueServiceCoordinationStore(
     @Override
     public CheckAndSetResult<ValueAndBound<T>> transformAgreedValue(Function<ValueAndBound<T>, T> transform) {
         Optional<SequenceAndBound> coordinationValue = getCoordinationValue();
-        T targetValue = transform.apply(
-                ValueAndBound.of(coordinationValue.flatMap(
-                        sequenceAndBound -> getValue(sequenceAndBound.sequence())),
-                        coordinationValue.map(SequenceAndBound::bound).orElse(SequenceAndBound.INVALID_BOUND)));
+        ValueAndBound<T> extantValueAndBound = ValueAndBound.of(coordinationValue.flatMap(
+                sequenceAndBound -> getValue(sequenceAndBound.sequence())),
+                coordinationValue.map(SequenceAndBound::bound).orElse(SequenceAndBound.INVALID_BOUND));
+        T targetValue = transform.apply(extantValueAndBound);
 
-        long sequenceNumber = sequenceNumberSupplier.getAsLong();
-        putUnlessValueExists(sequenceNumber, targetValue);
+        SequenceAndBound newSequenceAndBound
+                = determineNewSequenceAndBound(coordinationValue, extantValueAndBound, targetValue);
 
-        long newBound = getNewBound(sequenceNumber);
         CheckAndSetResult<SequenceAndBound> casResult = checkAndSetCoordinationValue(
-                coordinationValue, SequenceAndBound.of(sequenceNumber, newBound));
-        return extractRelevantValues(targetValue, newBound, casResult);
+                coordinationValue, newSequenceAndBound);
+        return extractRelevantValues(targetValue, newSequenceAndBound.bound(), casResult);
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType") // Passed from other operations returning Optional
+    private SequenceAndBound determineNewSequenceAndBound(
+            Optional<SequenceAndBound> coordinationValue,
+            ValueAndBound<T> extantValueAndBound,
+            T targetValue) {
+        long sequenceNumber;
+        long newBound;
+        if (shouldReuseExtantValue(coordinationValue, extantValueAndBound.value(), targetValue)) {
+            // Safe as we're only on this branch if the value is present
+            sequenceNumber = coordinationValue.get().sequence();
+            newBound = getNewBound(sequenceNumberSupplier.getAsLong());
```

Comment:
nit: maybe update the argument of `getNewBound` to not take a sequence number, since we're clearly not passing in the sequence number here. I guess this is just some blurring of sequence numbers and timestamps that's hard to resolve.

Reply:
Yep, that's a good call. I've updated the argument to `pointInTime` rather than `sequenceNumber`. Would it make sense to have a separate `boundSupplier` (potentially taking an extra timelock rpc?)

Reply:
I'm not too concerned, since this code won't get changed much (I hope), as long as it's clear to someone reading the code what we're doing and that it's correct - this line looked a bit fishy to me at first glance.

Reply:
OK. Yeah, `sequenceNumber` for that method is definitely misleading after this PR. I'll leave it as `pointInTime` for now.